File: frontend/lib/entitlements.ts
/**
 * Feature Entitlements — Decision-Maturity Segmentation
 *
 * Tier progression mirrors analytical capability growth:
 *   Starter  → Awareness Engine     (observe, orient)
 *   Investor → Decision Intelligence (decide, size)
 *   Trader   → Allocation & Execution Infrastructure (deploy, manage)
 *
 * Rules:
 * - Admins always have full access.
 * - A higher tier inherits all lower-tier features.
 * - Features should feel like expansion, not unlocking basics.
 */

export type Tier = 'starter' | 'investor' | 'trader'

export type Feature =
  | 'historical_patterns'       // Historical Analog Panel — pattern framing
  | 'institutional_risk'        // Institutional Risk Dashboard — stability / noise / risk efficiency
  | 'probabilistic_engine'      // Probabilistic Engine — full EV model, stop probability, scenario distribution
  | 'analyst_verdict'           // Full Analyst Verdict — investment thesis, full agents report
  | 'execution_layer'           // Execution Layer — position sizing, portfolio risk, trade stability
  | 'trade_setup_details'       // Trader-only: enhanced trade setup table + fund/tech divergence
  // ── Investor+ section gates (Discipline Layer) ───────────────────────────
  | 'sizing_summary'            // All tiers: allocation % + plain-language rationale
  | 'signal_metrics'            // Investor+: σ band, noise score, stop probability headline
  | 'stop_probability_detail'   // Investor+: stop prob decomposition table
  | 'full_scenario_table'       // Investor+: full bear/base/bull probability table
  | 'capital_deployment_path'   // Investor+: 3-stage tranche deployment path
  | 'thesis_break_monitor'      // Investor+: thesis break conditions monitor
  | 'structural_quality_full'   // Investor+: moat & quality score breakdown
  | 'stop_probability_basic'    // Investor+: stop probability headline value
  | 'structural_deployment_update' // Investor+: monthly capital deployability report
  // ── Trader-only section gates (Edge Layer) ───────────────────────────────
  | 'engine_diagnostics'        // Trader: full engine panels + driver ranking
  | 'scenario_weights'          // Trader: model vs effective scenario weights + rotation
  | 'multiplier_stack'          // Trader: position-sizing multiplier list + product
  | 'risk_matrix_full'          // Trader: full portfolio interaction metrics
  | 'ev_stability_panel'        // Trader: EV stability diagnostics (future)
  | 'probability_reliability'   // Trader: probability reliability scores (future)
  | 'stop_probability_advanced' // Trader: advanced stop probability decomposition (future)
  | 'universe_context'          // Trader: universe percentile context (future)
  | 'percentile_ranks'          // Trader: percentile rank overlays (future)
  | 'sensitivity_band'          // Trader: sensitivity band visualization (future)
  | 'horizon_normalization'     // Trader: horizon efficiency scores (future)

/** Minimum tier required to access each feature. */
const FEATURE_REQUIREMENTS: Record<Feature, Tier> = {
  // Pre-existing features
  historical_patterns:         'investor',
  institutional_risk:          'investor',
  probabilistic_engine:        'investor',
  analyst_verdict:             'investor',
  execution_layer:             'trader',
  trade_setup_details:         'trader',
  sizing_summary:              'starter',
  signal_metrics:              'investor',
  stop_probability_detail:     'investor',
  engine_diagnostics:          'trader',
  scenario_weights:            'trader',
  multiplier_stack:            'trader',
  risk_matrix_full:            'trader',
  structural_deployment_update: 'investor',
  // New Investor+ gates (Discipline Layer)
  full_scenario_table:         'investor',
  capital_deployment_path:     'investor',
  thesis_break_monitor:        'investor',
  structural_quality_full:     'investor',
  stop_probability_basic:      'investor',
  // New Trader-only gates (Edge Layer)
  ev_stability_panel:          'trader',
  probability_reliability:     'trader',
  stop_probability_advanced:   'trader',
  universe_context:            'trader',
  percentile_ranks:            'trader',
  sensitivity_band:            'trader',
  horizon_normalization:       'trader',
}

const TIER_ORDER: Record<Tier, number> = {
  starter:  0,
  investor: 1,
  trader:   2,
}

/** Returns true if `userTier` meets or exceeds the required tier for `feature`. */
export function canAccessFeature(feature: Feature, userTier: Tier | string | null | undefined): boolean {
  if (!userTier) return false
  const required = FEATURE_REQUIREMENTS[feature]
  const userLevel = TIER_ORDER[userTier as Tier] ?? -1
  const requiredLevel = TIER_ORDER[required]
  return userLevel >= requiredLevel
}

/** Display name for each tier. */
export const TIER_LABELS: Record<Tier, string> = {
  starter:  'Starter',
  investor: 'Investor',
  trader:   'Trader',
}

/** What each locked feature unlocks (for upgrade prompts). */
export const FEATURE_GATE_COPY: Record<Feature, {
  title: string
  description: string
  requiredTier: Tier
  bullets: string[]
}> = {
  historical_patterns: {
    title: 'Historical Pattern Framing',
    description: 'Structural context from analogous market regimes.',
    requiredTier: 'investor',
    bullets: [
      'Match current setup to historical analogs',
      'Understand base-rate outcomes for this pattern type',
      'Calibrate conviction with regime-level context',
    ],
  },
  institutional_risk: {
    title: 'Stability & Noise Diagnostics',
    description: 'Quantified signal reliability, factor exposure, and liquidity risk.',
    requiredTier: 'investor',
    bullets: [
      'Volatility regime & liquidity risk assessment',
      'Signal noise / reliability scoring',
      'Sensitivity attribution across risk factors',
      'Drift diagnostics — detect model instability early',
    ],
  },
  probabilistic_engine: {
    title: 'Probabilistic Engine & EV Model',
    description: 'Full expected value framework with scenario-weighted probabilities.',
    requiredTier: 'investor',
    bullets: [
      'Expected value across Bear / Base / Bull scenarios',
      'Stop probability framework & loss distribution',
      'Confidence interval calibration',
      'Longitudinal model drift comparison',
    ],
  },
  analyst_verdict: {
    title: 'Full Analyst Verdict',
    description: 'Structured investment thesis synthesised from all three agent reports.',
    requiredTier: 'investor',
    bullets: [
      'Company overview & recommendation summary',
      'Investment highlights with supporting data',
      'Key risks with specifics, not generics',
      'Entry strategy & investor fit guidance',
    ],
  },
  execution_layer: {
    title: 'Execution & Allocation Infrastructure',
    description: 'Capital deployment tools for position sizing and portfolio construction.',
    requiredTier: 'trader',
    bullets: [
      'Dynamic position sizing (noise-adjusted exposure)',
      'Portfolio risk engine & concentration diagnostics',
      'Factor exposure approximation',
      'Trade stability monitoring & correlation crowding',
    ],
  },
  trade_setup_details: {
    title: 'Trade Setup Details',
    description: 'Conservative vs aggressive entry/exit tables with 3-target playbook.',
    requiredTier: 'trader',
    bullets: [
      'Conservative and aggressive entry price levels',
      'Three-tier exit targets with partial-sell percentages',
      'Stop-loss levels and risk/reward ratios',
      'Fundamental vs technical divergence analysis',
    ],
  },
  sizing_summary: {
    title: 'Position Sizing Summary',
    description: 'Recommended allocation percentage and plain-language sizing rationale.',
    requiredTier: 'starter',
    bullets: [
      'Recommended portfolio allocation %',
      'Plain-language conviction rationale',
    ],
  },
  signal_metrics: {
    title: 'Signal Metrics',
    description: 'Numeric engine diagnostics: σ dispersion, noise score, and stop probability headline.',
    requiredTier: 'investor',
    bullets: [
      'EV sensitivity band (σ) and stability classification',
      'Noise score and noise regime label',
      'Stop probability headline value',
      'EV confidence level and dispersion',
    ],
  },
  stop_probability_detail: {
    title: 'Stop Probability Decomposition',
    description: 'Component-level breakdown of what drives your stop-out probability.',
    requiredTier: 'investor',
    bullets: [
      'Base stop risk from volatility regime',
      'Trend and support modifiers',
      'Effective stop probability with narrative',
    ],
  },
  engine_diagnostics: {
    title: 'Full Engine Diagnostics',
    description: 'Complete probabilistic engine output with driver ranking and attribution.',
    requiredTier: 'trader',
    bullets: [
      'EV sensitivity attribution by driver type',
      'Signal vs market movement decomposition',
      'Confidence integrity with degradation breakdown',
      'Full diagnostic panels for all engine modules',
    ],
  },
  scenario_weights: {
    title: 'Scenario Weight Diagnostics',
    description: 'Model vs effective scenario probabilities with rotation and compression analysis.',
    requiredTier: 'trader',
    bullets: [
      'Bear / Base / Bull model vs effective weights',
      'Scenario rotation index and compression ratio',
      'Tail state classification (Expanded / Neutral / Compressed)',
      'Active rotation factors driving weight shifts',
    ],
  },
  multiplier_stack: {
    title: 'Position Sizing Multiplier Stack',
    description: 'Full multiplier breakdown showing how each factor adjusts your position size.',
    requiredTier: 'trader',
    bullets: [
      'Noise, sensitivity, dispersion, and stop-risk multipliers',
      'Product of multipliers and net adjustment',
      'Cap state and constraint events',
      'Conviction justification for the final allocation',
    ],
  },
  risk_matrix_full: {
    title: 'Full Risk Matrix',
    description: 'Complete portfolio interaction metrics including correlation and crowding data.',
    requiredTier: 'trader',
    bullets: [
      'Correlation with existing portfolio positions',
      'Crowding and concentration risk flags',
      'Factor exposure breakdown',
      'Liquidity and volatility risk interaction metrics',
    ],
  },
  structural_deployment_update: {
    title: 'Structural Deployment Update',
    description: 'Monthly capital deployability report — structurally confirmed names only.',
    requiredTier: 'investor',
    bullets: [
      'Market deployability snapshot with capital posture classification',
      'Confirmed deployable tickers filtered by 5-criterion structural gate',
      'Sector breadth overview with trend direction',
    ],
  },
  // ── New Investor+ gates ────────────────────────────────────────────────
  full_scenario_table: {
    title: 'Full Scenario Probability Table',
    description: 'Bear / Base / Bull scenarios with probability weights and expected value.',
    requiredTier: 'investor',
    bullets: [
      'Probability-weighted expected value across all three scenarios',
      'Bear / Base / Bull price targets with supporting rationale',
      'Scenario rotation index and compression ratio',
    ],
  },
  capital_deployment_path: {
    title: 'Capital Deployment Path',
    description: '3-stage tranche deployment framework — when and how much to add.',
    requiredTier: 'investor',
    bullets: [
      '3-stage entry plan tied to price and confirmation signals',
      'Stage trigger conditions with live status tracking',
      'Max position sizing with structured add-on logic',
    ],
  },
  thesis_break_monitor: {
    title: 'Thesis Break Monitor',
    description: 'Conditions that would invalidate the investment thesis.',
    requiredTier: 'investor',
    bullets: [
      'Monitored thesis break conditions with live thresholds',
      'Automatic position sizing reduction on thesis breach',
      'Break severity classification (partial vs full exit)',
    ],
  },
  structural_quality_full: {
    title: 'Structural Quality Breakdown',
    description: 'Full moat, earnings durability, and financial health scoring.',
    requiredTier: 'investor',
    bullets: [
      'Component-level moat breakdown across 5 quality dimensions',
      'Earnings durability and financial health scores',
      'Overall quality score with conviction weighting',
    ],
  },
  stop_probability_basic: {
    title: 'Stop Probability',
    description: 'Headline probability of hitting your stop-loss before target.',
    requiredTier: 'investor',
    bullets: [
      'Stop probability headline from the probabilistic engine',
      'Volatility regime and trend modifier context',
      'Effective stop level with narrative explanation',
    ],
  },
  // ── New Trader-only gates ──────────────────────────────────────────────
  ev_stability_panel: {
    title: 'EV Stability Panel',
    description: 'Expected value stability diagnostics across market regimes.',
    requiredTier: 'trader',
    bullets: [
      'EV regime stability classification and drift index',
      'Sensitivity attribution by driver type',
      'Confidence integrity with degradation breakdown',
    ],
  },
  probability_reliability: {
    title: 'Probability Reliability Scores',
    description: 'Calibration quality of the probabilistic model for this setup.',
    requiredTier: 'trader',
    bullets: [
      'Model calibration score and historical accuracy',
      'Signal-to-noise ratio for this ticker and regime',
      'Reliability confidence interval',
    ],
  },
  stop_probability_advanced: {
    title: 'Advanced Stop Decomposition',
    description: 'Full driver-level breakdown of stop probability components.',
    requiredTier: 'trader',
    bullets: [
      'Base, trend, support, and volatility modifier breakdown',
      'Driver ranking by contribution to stop risk',
      'Scenario-conditional stop probabilities',
    ],
  },
  universe_context: {
    title: 'Universe Context',
    description: 'How this position compares to the full analysis universe.',
    requiredTier: 'trader',
    bullets: [
      'Percentile rank vs analyzed universe on quality and EV',
      'Sector and factor positioning context',
      'Relative crowding and concentration risk',
    ],
  },
  percentile_ranks: {
    title: 'Percentile Rank Overlays',
    description: 'Ranked positioning of key metrics vs historical and universe context.',
    requiredTier: 'trader',
    bullets: [
      'Quality score percentile vs sector and universe',
      'EV and stop probability percentile context',
      'Historical percentile for current valuation regime',
    ],
  },
  sensitivity_band: {
    title: 'Sensitivity Band',
    description: 'EV sensitivity range under assumption stress-testing.',
    requiredTier: 'trader',
    bullets: [
      'EV sensitivity band (σ) across key assumptions',
      'Upside/downside asymmetry under stressed inputs',
      'Stability classification with degradation flags',
    ],
  },
  horizon_normalization: {
    title: 'Horizon Efficiency',
    description: 'Return-per-unit-risk normalized across time horizons.',
    requiredTier: 'trader',
    bullets: [
      'Horizon-adjusted EV with time decay modeling',
      'Risk-efficiency rank vs typical setup duration',
      'Optimal hold period based on EV trajectory',
    ],
  },
}

File: api/lib/entitlements.py
"""
Backend entitlement feature flags for DVRG.

Feature matrix — Clarity Layer (Starter) / Discipline Layer (Investor) / Edge Layer (Trader):

  Core (all tiers):
    feature.report.core                  → Free, Starter, Investor, Trader
    feature.report.sizing_summary        → Free, Starter, Investor, Trader  (allocation % + rationale)

  Investor+ (Discipline Layer):
    feature.report.export_pdf            → Investor, Trader
    feature.report.signal_metrics        → Investor, Trader            (σ, noise score, stop prob headline)
    feature.report.stop_probability_detail → Investor, Trader          (decomposition table)
    feature.report.full_scenario_table   → Investor, Trader            (full bear/base/bull prob table)
    feature.report.capital_deployment_path → Investor, Trader          (3-stage tranche deployment)
    feature.report.thesis_break_monitor  → Investor, Trader            (thesis break section)
    feature.report.structural_quality_full → Investor, Trader          (moat + quality breakdown)
    feature.report.stop_probability_basic → Investor, Trader           (stop prob headline — alias)
    feature.deployment.structural_update → Investor, Trader

  Trader only (Edge Layer):
    feature.report.trade_setup           → Trader only
    feature.report.engine_diagnostics    → Trader only                 (full panels + driver ranking)
    feature.report.scenario_weights      → Trader only                 (model vs effective weights)
    feature.report.multiplier_stack      → Trader only                 (multiplier list + product)
    feature.report.risk_matrix_full      → Trader only                 (full portfolio interaction metrics)
    feature.report.ev_stability_panel    → Trader only                 (EV stability diagnostics)
    feature.report.probability_reliability → Trader only               (probability reliability scores)
    feature.report.stop_probability_advanced → Trader only             (advanced stop decomposition)
    feature.report.universe_context      → Trader only                 (universe percentile context)
    feature.report.percentile_ranks      → Trader only                 (percentile rank overlays)
    feature.report.sensitivity_band      → Trader only                 (sensitivity band visualization)
    feature.report.horizon_normalization → Trader only                 (horizon efficiency scores)

  Free tier has identical report content to Starter (same feature flags).
  The difference is how generation is gated: lifetime credits (2 total) vs. monthly quota.

Usage:
    from api.lib.entitlements import has_feature

    if not has_feature(user, FEAT_SIGNAL_METRICS):
        raise HTTPException(403, {"code": "NOT_ENTITLED", ...})
"""

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from api.models.auth import User

# ── Feature flag constants ─────────────────────────────────────────────────

# Core — all tiers
FEAT_REPORT_CORE            = "feature.report.core"
FEAT_SIZING_SUMMARY         = "feature.report.sizing_summary"

# Investor+ — Discipline Layer
FEAT_REPORT_PDF                  = "feature.report.export_pdf"
FEAT_SIGNAL_METRICS              = "feature.report.signal_metrics"
FEAT_STOP_PROB_DETAIL            = "feature.report.stop_probability_detail"
FEAT_FULL_SCENARIO_TABLE         = "feature.report.full_scenario_table"
FEAT_CAPITAL_DEPLOYMENT_PATH     = "feature.report.capital_deployment_path"
FEAT_THESIS_BREAK_MONITOR        = "feature.report.thesis_break_monitor"
FEAT_STRUCTURAL_QUALITY_FULL     = "feature.report.structural_quality_full"
FEAT_STOP_PROB_BASIC             = "feature.report.stop_probability_basic"
FEAT_DEPLOYMENT_STRUCTURAL       = "feature.deployment.structural_update"

# Trader only — Edge Layer
FEAT_REPORT_TRADE_SETUP          = "feature.report.trade_setup"
FEAT_ENGINE_DIAGNOSTICS          = "feature.report.engine_diagnostics"
FEAT_SCENARIO_WEIGHTS            = "feature.report.scenario_weights"
FEAT_MULTIPLIER_STACK            = "feature.report.multiplier_stack"
FEAT_RISK_MATRIX_FULL            = "feature.report.risk_matrix_full"
FEAT_EV_STABILITY_PANEL          = "feature.report.ev_stability_panel"
FEAT_PROBABILITY_RELIABILITY     = "feature.report.probability_reliability"
FEAT_STOP_PROB_ADVANCED          = "feature.report.stop_probability_advanced"
FEAT_UNIVERSE_CONTEXT            = "feature.report.universe_context"
FEAT_PERCENTILE_RANKS            = "feature.report.percentile_ranks"
FEAT_SENSITIVITY_BAND            = "feature.report.sensitivity_band"
FEAT_HORIZON_NORMALIZATION       = "feature.report.horizon_normalization"

# ── Ordered list of all flags (used by /api/entitlements response) ─────────
ALL_FEATURES: list[str] = [
    # Core
    FEAT_REPORT_CORE,
    FEAT_SIZING_SUMMARY,
    # Investor+
    FEAT_REPORT_PDF,
    FEAT_SIGNAL_METRICS,
    FEAT_STOP_PROB_DETAIL,
    FEAT_FULL_SCENARIO_TABLE,
    FEAT_CAPITAL_DEPLOYMENT_PATH,
    FEAT_THESIS_BREAK_MONITOR,
    FEAT_STRUCTURAL_QUALITY_FULL,
    FEAT_STOP_PROB_BASIC,
    FEAT_DEPLOYMENT_STRUCTURAL,
    # Trader only
    FEAT_REPORT_TRADE_SETUP,
    FEAT_ENGINE_DIAGNOSTICS,
    FEAT_SCENARIO_WEIGHTS,
    FEAT_MULTIPLIER_STACK,
    FEAT_RISK_MATRIX_FULL,
    FEAT_EV_STABILITY_PANEL,
    FEAT_PROBABILITY_RELIABILITY,
    FEAT_STOP_PROB_ADVANCED,
    FEAT_UNIVERSE_CONTEXT,
    FEAT_PERCENTILE_RANKS,
    FEAT_SENSITIVITY_BAND,
    FEAT_HORIZON_NORMALIZATION,
]

# ── Tier entitlement matrix ────────────────────────────────────────────────
_INVESTOR_PLUS: set[str] = {
    FEAT_REPORT_PDF,
    FEAT_SIGNAL_METRICS,
    FEAT_STOP_PROB_DETAIL,
    FEAT_FULL_SCENARIO_TABLE,
    FEAT_CAPITAL_DEPLOYMENT_PATH,
    FEAT_THESIS_BREAK_MONITOR,
    FEAT_STRUCTURAL_QUALITY_FULL,
    FEAT_STOP_PROB_BASIC,
    FEAT_DEPLOYMENT_STRUCTURAL,
}

_TRADER_ONLY: set[str] = {
    FEAT_REPORT_TRADE_SETUP,
    FEAT_ENGINE_DIAGNOSTICS,
    FEAT_SCENARIO_WEIGHTS,
    FEAT_MULTIPLIER_STACK,
    FEAT_RISK_MATRIX_FULL,
    FEAT_EV_STABILITY_PANEL,
    FEAT_PROBABILITY_RELIABILITY,
    FEAT_STOP_PROB_ADVANCED,
    FEAT_UNIVERSE_CONTEXT,
    FEAT_PERCENTILE_RANKS,
    FEAT_SENSITIVITY_BAND,
    FEAT_HORIZON_NORMALIZATION,
}

_CORE: set[str] = {FEAT_REPORT_CORE, FEAT_SIZING_SUMMARY}

_ENTITLEMENTS: dict[str, set[str]] = {
    # Free tier: identical content to Starter (gating is credit-based, not feature-based)
    "free":     _CORE.copy(),
    "starter":  _CORE.copy(),
    "investor": _CORE | _INVESTOR_PLUS,
    "trader":   _CORE | _INVESTOR_PLUS | _TRADER_ONLY,
}

# Stripe statuses that revoke paid-tier features
_INACTIVE_STATUSES: frozenset[str] = frozenset({
    "canceled", "past_due", "unpaid", "incomplete_expired",
})


def _effective_tier(user: "User") -> str:
    """
    Return the tier that should be used for entitlement checks.

    Admins are never downgraded.
    Users whose Stripe subscription is in an inactive status are treated as
    Starter for entitlement purposes (no grace period here; implement above if
    needed).
    """
    if user.is_admin:
        return "trader"  # admins get everything

    tier = str(user.tier.value).lower() if hasattr(user.tier, "value") else str(user.tier).lower()

    stripe_status = user.stripe_subscription_status
    # Inactive Stripe subscription: paid users drop to starter, free users stay free
    if stripe_status and stripe_status in _INACTIVE_STATUSES:
        return "free" if tier == "free" else "starter"

    return tier


def has_feature(user: "User", feature: str) -> bool:
    """
    Return True if the user is entitled to the given feature flag.

    Args:
        user: Authenticated User model instance.
        feature: One of the FEAT_* constants defined in this module.
    """
    tier = _effective_tier(user)
    return feature in _ENTITLEMENTS.get(tier, _ENTITLEMENTS["free"])


def upgrade_hint(feature: str) -> str:
    """Return the minimum tier name needed to access a feature."""
    for tier in ("starter", "investor", "trader"):
        if feature in _ENTITLEMENTS.get(tier, set()):
            return tier
    return "trader"
